### PR TITLE
substream(): ensure identical behavior to read_bytes()

### PR DESCRIPTION
**Pull request by generalmimon**
Fixes the `{Eof,Eos}ExceptionSized` tests in Ruby (see https://github.com/kaitai-io/ci_artifacts/blob/d7568fe794b09292b98873dcdc1a5534928345b6/test_out/ruby/ci.json#L314-L324)

Overall, this PR tries to ensure that the zero-copy substreams (https://github.com/kaitai-io/kaitai_struct/issues/44) are a drop-in replacement for the old way using `read_bytes` + `Stream#new` without any observable differences. It's not perfect, but it should be close enough.

To verify this, we automatically generate and evaluate 2000 random scenarios using https://github.com/rantly-rb/rantly involving 10 stream operations and compare whether the old and new substream implementations behave the same. If not, the test will fail. It is a normal RSpec test, so it can be run along with other unit tests using `rspec spec/` on the command line.

Every single change made to the stream implementation follows from a particular edge case identified by this extensive testing.

### CODE REVIEW

**Comment by generalmimon** [comment]
@GreyCat You can take a look. I want to have this merged before releasing a new version, because the failure of `{Eof,Eos}ExceptionSized` tests (see https://github.com/kaitai-io/kaitai_struct_tests/commit/99f9229714a29dc8353efecf5749688c0397b4ef) visible at https://ci.kaitai.io/ means that this kind of EOF errors would not be reported (or detected) with zero-copy substreams, which is a regression since 0.10 because 0.10 used the old `read_bytes` + `Stream#new` way to do substreams, but right now zero-copy substreams are enabled by default.

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+group :test do
+  gem 'rantly', '~> 3.0'
+  gem 'rspec', '~> 3.13'
+end
+
+group :development do
+  gem 'rubocop', require: false
+  gem 'rubocop-rspec', require: false
+end
```
Comment (GreyCat) [inline Gemfile:13]: I might be missing something, but [according to this rather old article](https://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/), even if Gemfile for a library which gets to be built as a gem itself is to be used, it should be just:

```ruby
source 'https://rubygems.org'

gemspec
```

Would that make sense to follow this approach, or you see some contradiction with that?

Comment (generalmimon) [reply Gemfile:13]: There are different opinions on that, but here I'm kind of following what I did in KSV (see https://github.com/kaitai-io/kaitai_struct_visualizer/commit/f49a18c30b035d8f14a4fbdd71157e8317169f3d), i.e. runtime dependencies to `.gemspec` (but the Ruby runtime doesn't have any at the moment), development dependencies to `Gemfile`. See https://github.com/rubygems/rubygems/issues/1104 for a bunch of rationales both in favor or against this.

Frankly, I don't care as long as it works.

Comment (GreyCat) [reply Gemfile:13]: Got it, thanks!

File: lib/kaitai/struct/struct.rb [hunk 2]
```diff
@@ -587,6 +608,29 @@ def to_signed(x, mask)
     (x & ~mask) - (x & mask)
   end
 
+  ##
+  # Internal implementation helper. Do not use from outside the
+  # library, it can be removed at any time.
```
Comment (GreyCat) [inline lib/kaitai/struct/struct.rb:613]: Technically, it's already declared as "private", and "don't use outside, can be removed at any time" is exactly the definition of private, and equally applies to all other private members. This one is not anyhow different from `to_signed()`, `format_hex()`, etc.

Comment (generalmimon) [reply lib/kaitai/struct/struct.rb:613]: > Technically, it's already declared as "private"

Unfortunately, that's not the case, as RuboCop points out:

```console
pp@DESKTOP-89OPGF3 MINGW64 /c/temp/kaitai_struct/runtime/ruby (fix-zero-copy-substreams)
$ rubocop --fail-level W --display-only-fail-level-offenses
Inspecting 5 files
.WWW.

Offenses:

(...)
lib/kaitai/struct/struct.rb:603:3: W: Lint/UselessConstantScoping: Useless private access modifier for constant scope.
  SIGN_MASK_16 = (1 << (16 - 1))
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/kaitai/struct/struct.rb:604:3: W: Lint/UselessConstantScoping: Useless private access modifier for constant scope.
  SIGN_MASK_32 = (1 << (32 - 1))
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/kaitai/struct/struct.rb:605:3: W: Lint/UselessConstantScoping: Useless private access modifier for constant scope.
  SIGN_MASK_64 = (1 << (64 - 1))
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/kaitai/struct/struct.rb:617:3: W: Lint/IneffectiveAccessModifier: private (on line 602) does not make singleton methods private. Use private_class_method or private inside a class << self block instead.
  def self.num2long(val)
  ^^^
lib/kaitai/struct/struct.rb:634:3: W: Lint/IneffectiveAccessModifier: private (on line 602) does not make singleton methods private. Use private_class_method or private inside a class << self block instead.
  def self.format_hex(bytes)
  ^^^
lib/kaitai/struct/struct.rb:649:3: W: Lint/IneffectiveAccessModifier: private (on line 602) does not make singleton methods private. Use private_class_method or private inside a class << self block instead.
  def self.is_byte_array?(*args)
  ^^^
lib/kaitai/struct/struct.rb:653:3: W: Lint/IneffectiveAccessModifier: private (on line 602) does not make singleton methods private. Use private_class_method or private inside a class << self block instead.
  def self.inspect_values(*args)
  ^^^
(...)

5 files inspected, 25 offenses detected, 12 offenses autocorrectable
```

It's true that `to_signed()` and `format_hex()` are no different, so they are in fact public as well.

Comment (generalmimon) [reply lib/kaitai/struct/struct.rb:613]: I don't think it's actually possible to make `num2long` private in Ruby in this case, because it is used in both `Kaitai::Struct::Stream#substream`:

https://github.com/kaitai-io/kaitai_struct_ruby_runtime/blob/e7a9f98c6ec7efd71be5b98bb7cf6ea61b2e792a/lib/kaitai/struct/struct.rb#L579-L582

... and `Kaitai::Struct::SubIO#seek`:

https://github.com/kaitai-io/kaitai_struct_ruby_runtime/blob/e7a9f98c6ec7efd71be5b98bb7cf6ea61b2e792a/lib/kaitai/struct/struct.rb#L703-L706

`Kaitai::Struct::SubIO` is a different class than `Kaitai::Struct::Stream`, so the usage of `Kaitai::Struct::Stream.num2long` from `Kaitai::Struct::SubIO#seek` requires public access. As far as I know, Ruby doesn't have a concept of _package-private_ like for example Java does.

Comment (GreyCat) [reply lib/kaitai/struct/struct.rb:613]: Got it, thanks for the full explanation — looks like my Ruby-fu has much room for improvement!

Scanning a few things around, it looks like there's indeed no close equivalent to package private (or module private) in Ruby, but it looks like standard practice for many gems would be moving these internal implementation details into a separate module = namespace, clearly named `Internal`, e.g.:

* https://github.com/mina-deploy/mina/blob/dc9deda3d678eb4702bc020b1ddd5a7d3456834b/lib/mina/helpers/internal.rb#L5
* https://github.com/thirdside/berlin-ai/blob/19dda95a0935f9fa4750e4bac8576b69e2959e61/lib/ai/node_internal.rb#L4

Given we've accumulated quite a bit of these "helper" methods, might worth doing that — at least we can get away with putting that "internal only" warning once on whole module rather than copying it to every single method.

Comment (generalmimon) [reply lib/kaitai/struct/struct.rb]: > but it looks like standard practice for many gems would be moving these internal implementation details into a separate module = namespace, clearly named `Internal`

Thanks for this suggestion, I did that in https://github.com/kaitai-io/kaitai_struct_ruby_runtime/pull/12/commits/5178ff7093eeba25baef3b56986b8d746ed59fda. Surprisingly, I was actually able to use `private_constant :Internal`, which makes the entire module private, so accessing it as `Kaitai::Struct::Internal` gives an error:

```console
pp@DESKTOP-89OPGF3 MINGW64 /c/temp/kaitai_struct/runtime/ruby (fix-zero-copy-substreams)
$ cat t.rb
require_relative 'lib/kaitai/struct/struct'

pp Kaitai::Struct::Internal.num2long(4.8)

pp@DESKTOP-89OPGF3 MINGW64 /c/temp/kaitai_struct/runtime/ruby (fix-zero-copy-substreams)
$ ruby t.rb
t.rb:3:in '<main>': private constant Kaitai::Struct::Internal referenced (NameError)

pp Kaitai::Struct::Internal.num2long(4.8)
                 ^^^^^^^^^^
```

However, since both `Kaitai::Struct::Stream#substream` and `Kaitai::Struct::SubIO#seek` see the `Internal` module directly as a "local variable", they can just call `Internal.num2long(n)` without issues. The important part is that they don't need to refer to the module via `Kaitai::Struct`, because they cannot do `Kaitai::Struct::Internal` either, since that would also count as an attempt to access the private `Internal` constant from the outside. But once they have a reference to `Internal`, it's all good because `num2long` is a _public_ method from the perspective of `Internal`.

I think that's pretty neat, and there we have the answer to the question of how to do package-private in Ruby.

Comment (GreyCat) [reply lib/kaitai/struct/struct.rb]: Sounds great! Thanks for checking and going forward with it!

Looks like I've learned something new today too :)

**Review by GreyCat — APPROVED** [review]
Looks good to me, thanks for contributing — these 2000 test cases permutations looks like really exhaustive testing coverage!

Three propositions I have for further steps (if you want, I can do these):

* Add README.md entry on how to do install / run tests for development purposes
* Add GH Action to do run the tests on every PR
* Add GH Action to do official build/release

**Comment by generalmimon** [comment]
Thanks for reviewing!
